File: src/pixel_perversion_engine/render/spine.clj
(ns pixel-perversion-engine.render.spine
  (:import [com.badlogic.gdx Gdx]
           [com.badlogic.gdx.graphics GL20 Pixmap Pixmap$Format]
           [com.badlogic.gdx.graphics.glutils FrameBuffer]
           [com.pixel_perversion_engine.render Render]
           [com.badlogic.gdx.utils.viewport Viewport]
           [com.badlogic.gdx.graphics.g2d TextureRegion]))

(def fbo nil)
(def fbo-region nil)

(defn setup-render
  []
  (def fbo (new FrameBuffer Pixmap$Format/RGBA8888 800 480 false))
  (def fbo-region (new TextureRegion (.getColorBufferTexture fbo) 0 0 800 480)))
(defn dispose-render
  []
  (.dispose fbo)
  (.dispose fbo-region))

(defn render
  [^Render render ^Viewport viewport objs shader-program]
  (let [
        camera (.getCamera viewport)
        sprite-batch (.getSpriteBatch render)
        polygon-batch (.getPolygonBatch render)
        skeleton-renderer (.getSkeletonRenderer render)]

    (.begin fbo)
    (.glClearColor Gdx/gl 0 0 0 0)
    (.glClear Gdx/gl GL20/GL_COLOR_BUFFER_BIT)

    (.apply viewport false)
    ;(.update camera)

    (.setShader polygon-batch nil) ;We normally do not set the shader for the polygon batch! This happens below within the sprite batch.
    (.setProjectionMatrix polygon-batch (.-combined camera))

    (.begin polygon-batch)

    (doseq [obj objs]
      (.draw skeleton-renderer polygon-batch (.getSkeleton (:renderable obj))))

    (.end polygon-batch)
    (.end fbo)

    (let [
          cam-x (.x (.position camera))
          cam-y (.y (.position camera))
          x (- cam-x (/ (.viewportWidth camera) 2))
          y (- cam-y (/ (.viewportHeight camera) 2))
          ]
      (.setTexture fbo-region (.getColorBufferTexture fbo))
      (if (not (.isFlipY fbo-region)) (.flip fbo-region false true))
      (.setProjectionMatrix sprite-batch (.-combined camera))
      (.begin sprite-batch)
      (.setShader sprite-batch shader-program)
      (.draw sprite-batch fbo-region (float x) (float y))
      (.end sprite-batch)
      )

    ;(.dispose fbo)
    ;(.dispose fbo-region)
    ))

File: src/src_java/com/pixel_perversion_engine/render/Spine.java
package com.pixel_perversion_engine.render;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.Pixmap;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.glutils.FrameBuffer;
import com.badlogic.gdx.graphics.glutils.ShaderProgram;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.viewport.Viewport;

import java.util.HashMap;

/**
 * Created by Travis on 11/21/2015.
 *
 */
public class Spine {
    static public void draw(
            Render render, Viewport viewport,
            Array<HashMap> spines, ShaderProgram shaderProgram){ //com.pixel_perversion_engine.spine.Spine
        //viewport.apply();

        FrameBuffer fbo = new FrameBuffer(Pixmap.Format.RGBA8888,
                800, 480, false);

        fbo.begin();
        Gdx.gl.glClearColor(0f, 0f, 0f, 0f);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        com.badlogic.gdx.graphics.Camera camera = viewport.getCamera();
        camera.update();

        //fully render spine object BEFORE applying shader!
        render.getPolygonBatch().setShader(null);
        render.getPolygonBatch().setProjectionMatrix(camera.combined);

        //draw spine object

        render.getPolygonBatch().begin();

        for (HashMap map : spines) {
            com.pixel_perversion_engine.spine.Spine spine = (com.pixel_perversion_engine.spine.Spine) map.get("renderable");
            render.getSkeletonRenderer().draw(render.getPolygonBatch(), spine.getSkeleton());
        }

        render.getPolygonBatch().end();

        fbo.end();

        TextureRegion fboRegion = new TextureRegion(fbo.getColorBufferTexture(), 0, 0, 800, 480);
        fboRegion.flip(false, true);

        //render.getSpriteBatch().setBlendFunction(GL20.GL_ONE, GL20.GL_ONE_MINUS_SRC_ALPHA);
        render.getSpriteBatch().begin();
        render.getSpriteBatch().setShader(shaderProgram);
        //render.getSpriteBatch().setColor(1f, 1f, 1f, 1f);
        //if (useShader) render.getSpriteBatch().setShader(shaderProgram);
        render.getSpriteBatch().draw(fboRegion, 0, 0);
        render.getSpriteBatch().end();

        fbo.dispose();
        //return fbo;
    }
}
